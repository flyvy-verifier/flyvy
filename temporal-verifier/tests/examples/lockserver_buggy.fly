# Copyright 2022-2023 VMware, Inc.
# SPDX-License-Identifier: BSD-2-Clause

# TEST -- bdd-check --bound node=3 --print-timing false
# TEST -- sat-check --bound node=3 --print-timing false --depth=20
# TEST -- set-check --bound node=3 --print-timing false
# TEST -- smt-check --print-timing false --depth=10

# To exploit this bug into a safety violation, you first need to have one node acquire and
# release the lock via the following 5 transitions:
# - send lock_msg, send grant, receive grant, send unlock, receive unlock
# then you need two nodes to both acquire the lock:
# - send lock_msg, send grant, receive grant (3 transitions)
# - receive unlock (1 transition, exploiting the fact that it was not removed)
# - send lock_msg, send grant, receive grant (3 transitions)
#
# In total, that's 5 + 3 + 1 + 3 = 12 transitions to exploit it.

sort node

mutable lock_msg(node): bool
mutable grant_msg(node): bool
mutable unlock_msg(node): bool
mutable holds_lock(node): bool
mutable server_holds_lock: bool

# inits:
assume (forall N:node. !lock_msg(N)) & (forall N:node. !grant_msg(N)) & (forall N:node.
    !unlock_msg(N)) & (forall N:node. !holds_lock(N)) & (server_holds_lock)

# transitions:
assume always
    (exists n:node. 
        (forall N:node. ((lock_msg(N))') <-> lock_msg(N) | N = n) & 
        (forall x0:node. ((grant_msg(x0))') = grant_msg(x0)) & 
        (forall x0:node. ((unlock_msg(x0))') = unlock_msg(x0)) & 
        (forall x0:node. ((holds_lock(x0))') = holds_lock(x0)) & 
        ((server_holds_lock)') = server_holds_lock) | 
    (exists n:node. 
        (forall N:node. server_holds_lock & lock_msg(n) & 
            !((server_holds_lock)') & 
            (((lock_msg(N))') <-> lock_msg(N) & N != n) & 
            (((grant_msg(N))') <-> grant_msg(N) | N = n)) & 
        (forall x0:node. ((unlock_msg(x0))') = unlock_msg(x0)) & 
        (forall x0:node. ((holds_lock(x0))') = holds_lock(x0))) | 
    (exists n:node. 
        (forall N:node. grant_msg(n) & 
            (((grant_msg(N))') <-> grant_msg(N) & N != n) & 
            (((holds_lock(N))') <-> holds_lock(N) | N = n)) & 
        (forall x0:node. ((lock_msg(x0))') = lock_msg(x0)) & 
        (forall x0:node. 
            ((unlock_msg(x0))') = unlock_msg(x0)) & 
            ((server_holds_lock)') = server_holds_lock) | 
    (exists n:node. 
        (forall N:node. holds_lock(n) & 
            (((holds_lock(N))') <-> holds_lock(N) & N != n) & 
            (((unlock_msg(N))') <-> unlock_msg(N) | N = n)) & 
        (forall x0:node. ((lock_msg(x0))') = lock_msg(x0)) &
        (forall x0:node. 
            ((grant_msg(x0))') = grant_msg(x0)) & 
            ((server_holds_lock)') = server_holds_lock) | 
    (exists n:node. 
        (forall N:node. unlock_msg(n) & 
            (((unlock_msg(N))') <-> unlock_msg(N)) & 
            ((server_holds_lock)')) & 
        (forall x0:node. ((lock_msg(x0))') = lock_msg(x0)) & 
        (forall x0:node. ((grant_msg(x0))') = grant_msg(x0)) & 
        (forall x0:node. ((holds_lock(x0))') = holds_lock(x0)))

# safety:
assert always (forall N1:node, N2:node. holds_lock(N1) & holds_lock(N2) -> N1 = N2)
