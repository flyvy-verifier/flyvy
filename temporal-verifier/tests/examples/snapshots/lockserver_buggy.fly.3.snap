---
source: temporal-verifier/tests/test_examples.rs
description: "--name=3 -- smt-check --print-timing false '--depth=10' tests/examples/lockserver_buggy.fly"
expression: combined_stdout_stderr
---
starting search...
answer: safe up to depth 10 for given sort bounds

======== STDERR: ===========

